src/proofs/inner_product.rs

```diff
@@ -218,6 +236,99 @@ impl InnerProductArg {
             Err(InnerProductError)
         }
     }
+
+    ///
+    /// Returns Ok() if the given inner product satisfies the verification equations,
+    /// else returns `InnerProductError`.
+    /// 
+    /// Uses a single multiexponentiation (multiscalar multiplication in additive notation)
+    /// check to verify an inner product proof. 
+    /// 
+    pub fn fast_verify(&self, g_vec: &[GE], hi_tag: &[GE], ux: &GE, P: &GE)
+    -> Result<(), Errors> {
+        let G = &g_vec[..];
+        let H = &hi_tag[..];
+        let n = G.len();
+        let order = FE::q();
+
+        // All of the input vectors must have the same length.
+        assert_eq!(G.len(), n);
+        assert_eq!(H.len(), n);
+        assert!(n.is_power_of_two());
+
+        let lg_n = self.L.len();
+
+        let mut x_sq_vec: Vec<BigInt> = Vec::with_capacity(lg_n);
+        let mut x_inv_sq_vec: Vec<BigInt> = Vec::with_capacity(lg_n);
+        let mut minus_x_sq_vec: Vec<BigInt> = Vec::with_capacity(lg_n);
+        let mut minus_x_inv_sq_vec: Vec<BigInt> = Vec::with_capacity(lg_n);
+        let mut allinv = BigInt::one();
+        for (Li, Ri) in self.L.iter().zip(self.R.iter()) {
+
+            let x = HSha256::create_hash_from_ge(&[&Li, &Ri, &ux]);
+            let x_bn = x.to_big_int();
+            let x_inv_fe = x.invert();
+            let x_inv_bn = x_inv_fe.to_big_int();
+            let x_sq_bn = BigInt::mod_mul(&x_bn, &x_bn, &order);
+            let x_inv_sq_bn =
+                BigInt::mod_mul(&x_inv_fe.to_big_int(), &x_inv_fe.to_big_int(), &order);
+            // let x_sq_fe: FE = ECScalar::from(&x_sq_bn);
+            // let x_inv_sq_fe: FE = ECScalar::from(&x_inv_sq_bn);
+
+            x_sq_vec.push(x_sq_bn.clone());
+            x_inv_sq_vec.push(x_inv_sq_bn.clone());
+            minus_x_sq_vec.push(BigInt::mod_sub(&BigInt::zero(), &x_sq_bn, &order));
+            minus_x_inv_sq_vec.push(BigInt::mod_sub(&BigInt::zero(), &x_inv_sq_bn, &order));
+            allinv = allinv * x_inv_bn;
+        }
+
+        let mut s: Vec<BigInt> = Vec::with_capacity(n);
+        s.push(allinv);
+        for i in 1..n {
+            let lg_i = (32 - 1 - (i as u32).leading_zeros()) as usize;
+            let k = 1 << lg_i;
+            // The challenges are stored in "creation order" as [x_k,...,x_1],
+            // so u_{lg(i)+1} = is indexed by (lg_n-1) - lg_i
+            let x_lg_i_sq = x_sq_vec[(lg_n - 1) - lg_i].clone();
+            s.push(s[i - k].clone() * x_lg_i_sq);
+        }
```

Comment:
could you explain in short what is the logic of computing the `s` vector? 
it is a bit hard to follow

Reply:
## A bit of theory

In the usual inner product proof verification, in the _n_-th round, we compute new base vectors computed as follows: 

<a href="https://www.codecogs.com/eqnedit.php?latex=\inline&space;\begin{align*}&space;\textbf{g}^{\prime}&space;&=&space;\textbf{g}^{x^{-1}}_{[:n^{\prime}]}&space;\&space;\circ&space;\&space;\textbf{g}^{x}_{[n^{\prime}:]}&space;\in&space;\mathbb{G}^{\frac{n}{2}}&space;\\&space;\textbf{h}^{\prime}&space;&=&space;\textbf{h}^{x}_{[:n^{\prime}]}&space;\&space;\circ&space;\&space;\textbf{h}^{x^{-1}}_{[n^{\prime}:]}&space;\in&space;\mathbb{G}^{\frac{n}{2}}&space;\end{align*}" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\inline&space;\begin{align*}&space;\textbf{g}^{\prime}&space;&=&space;\textbf{g}^{x^{-1}}_{[:n^{\prime}]}&space;\&space;\circ&space;\&space;\textbf{g}^{x}_{[n^{\prime}:]}&space;\in&space;\mathbb{G}^{\frac{n}{2}}&space;\\&space;\textbf{h}^{\prime}&space;&=&space;\textbf{h}^{x}_{[:n^{\prime}]}&space;\&space;\circ&space;\&space;\textbf{h}^{x^{-1}}_{[n^{\prime}:]}&space;\in&space;\mathbb{G}^{\frac{n}{2}}&space;\end{align*}" title="\begin{align*} \textbf{g}^{\prime} &= \textbf{g}^{x^{-1}}_{[:n^{\prime}]} \ \circ \ \textbf{g}^{x}_{[n^{\prime}:]} \in \mathbb{G}^{\frac{n}{2}} \\ \textbf{h}^{\prime} &= \textbf{h}^{x}_{[:n^{\prime}]} \ \circ \ \textbf{h}^{x^{-1}}_{[n^{\prime}:]} \in \mathbb{G}^{\frac{n}{2}} \end{align*}" /></a>

Thus, the base vectors used in each round depend only on the challenges  <a href="https://www.codecogs.com/eqnedit.php?latex=x_1,&space;x_2,&space;\dots,&space;x_{\lceil&space;\text{log}_2n&space;\rceil}" target="_blank"><img src="https://latex.codecogs.com/gif.latex?x_1,&space;x_2,&space;\dots,&space;x_{\lceil&space;\text{log}_2n&space;\rceil}" title="x_1, x_2, \dots, x_{\lceil \text{log}_2n \rceil}" /></a> used in each round. In _i_-th round, the verifier computes <a href="https://www.codecogs.com/eqnedit.php?latex=P_{i&plus;1}" target="_blank"><img src="https://latex.codecogs.com/gif.latex?P_{i&plus;1}" title="P_{i+1}" /></a> which would be used in the next round. It is computed as:

<a href="https://www.codecogs.com/eqnedit.php?latex=P_{i&plus;1}&space;=&space;L_{i}^{x_i^{2}}&space;P_{i}&space;R_{i}^{x_i^{-2}}" target="_blank"><img src="https://latex.codecogs.com/gif.latex?P_{i&plus;1}&space;=&space;L_{i}^{x_i^{2}}&space;P_{i}&space;R_{i}^{x_i^{-2}}" title="P_{i+1} = L_{i}^{x_i^{2}} P_{i} R_{i}^{x_i^{-2}}" /></a>

In the last round, the verifier checks the following equality:

<a href="https://www.codecogs.com/eqnedit.php?latex=P_{\lceil&space;\text{log}_2n&space;\rceil&space;&plus;&space;1}&space;=&space;g^a&space;h^bu^{a&space;\cdot&space;b}" target="_blank"><img src="https://latex.codecogs.com/gif.latex?P_{\lceil&space;\text{log}_2n&space;\rceil&space;&plus;&space;1}&space;=&space;g^a&space;h^bu^{a&space;\cdot&space;b}" title="P_{\lceil \text{log}_2n \rceil + 1} = g^a h^bu^{a \cdot b}" /></a>

where <a href="https://www.codecogs.com/eqnedit.php?latex=g,h" target="_blank"><img src="https://latex.codecogs.com/gif.latex?g,h" title="g,h" /></a> are generators used in the last round of the protocol.

For a single multi-exponentiation verification, a verifier just needs to just check the last verification equation. For that, a verifier will need the following quantities:

<a href="https://www.codecogs.com/eqnedit.php?latex=g,h,&space;\text{&space;and&space;}&space;P_{\lceil&space;\text{log}_2n&space;\rceil&space;&plus;&space;1}" target="_blank"><img src="https://latex.codecogs.com/gif.latex?g,h,&space;\text{&space;and&space;}&space;P_{\lceil&space;\text{log}_2n&space;\rceil&space;&plus;&space;1}" title="g,h, \text{ and } P_{\lceil \text{log}_2n \rceil + 1}" /></a>

Computing <a href="https://www.codecogs.com/eqnedit.php?latex=P_{\lceil&space;\text{log}_2n&space;\rceil&space;&plus;&space;1}" target="_blank"><img src="https://latex.codecogs.com/gif.latex?P_{\lceil&space;\text{log}_2n&space;\rceil&space;&plus;&space;1}" title="P_{\lceil \text{log}_2n \rceil + 1}" /></a> is easy. We need to do the following:

<a href="https://www.codecogs.com/eqnedit.php?latex=\begin{align*}&space;P_{\lceil&space;\text{log}_2n&space;\rceil&space;&plus;&space;1}&space;&=&space;L_{\lceil&space;\text{log}_2n&space;\rceil}^{(x_j^2)}&space;\cdot&space;P_{\lceil&space;\text{log}_2n&space;\rceil}&space;\cdot&space;R_{\lceil&space;\text{log}_2n&space;\rceil}^{(x_j^{-2})}&space;\\&space;&=&space;L_{\lceil&space;\text{log}_2n&space;\rceil}^{(x_j^2)}&space;\cdot&space;\left(&space;L_{\lceil&space;\text{log}_2n&space;\rceil&space;-1}^{(x_j^2)}&space;\cdot&space;P_{\lceil&space;\text{log}_2n&space;\rceil&space;-1}&space;\cdot&space;R_{\lceil&space;\text{log}_2n&space;\rceil&space;-1}^{(x_j^{-2})}&space;\right)&space;\cdot&space;R_{\lceil&space;\text{log}_2n&space;\rceil}^{(x_j^{-2})}&space;\\&space;&=&space;L_{\lceil&space;\text{log}_2n&space;\rceil}^{(x_j^2)}&space;\cdot&space;L_{\lceil&space;\text{log}_2n&space;\rceil&space;-1}^{(x_j^2)}&space;\bigg(&space;\&space;.&space;\&space;.&space;\&space;.&space;\&space;.&space;\&space;.&space;\&space;.&space;\&space;\bigg)&space;R_{\lceil&space;\text{log}_2n&space;\rceil&space;-1}^{(x_j^{-2})}&space;\cdot&space;R_{\lceil&space;\text{log}_2n&space;\rceil}^{(x_j^{-2})}&space;\\&space;&=&space;\left(&space;\prod_{j=1}^{\lceil&space;\text{log}_2n&space;\rceil}&space;L_{j}^{(x_j^2)}&space;\right)&space;\cdot&space;P&space;\cdot&space;\left(&space;\prod_{j=1}^{\lceil&space;\text{log}_2n&space;\rceil}&space;R_{j}^{(x_j^{-2})}&space;\right)&space;\end{align}" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\begin{align*}&space;P_{\lceil&space;\text{log}_2n&space;\rceil&space;&plus;&space;1}&space;&=&space;L_{\lceil&space;\text{log}_2n&space;\rceil}^{(x_j^2)}&space;\cdot&space;P_{\lceil&space;\text{log}_2n&space;\rceil}&space;\cdot&space;R_{\lceil&space;\text{log}_2n&space;\rceil}^{(x_j^{-2})}&space;\\&space;&=&space;L_{\lceil&space;\text{log}_2n&space;\rceil}^{(x_j^2)}&space;\cdot&space;\left(&space;L_{\lceil&space;\text{log}_2n&space;\rceil&space;-1}^{(x_j^2)}&space;\cdot&space;P_{\lceil&space;\text{log}_2n&space;\rceil&space;-1}&space;\cdot&space;R_{\lceil&space;\text{log}_2n&space;\rceil&space;-1}^{(x_j^{-2})}&space;\right)&space;\cdot&space;R_{\lceil&space;\text{log}_2n&space;\rceil}^{(x_j^{-2})}&space;\\&space;&=&space;L_{\lceil&space;\text{log}_2n&space;\rceil}^{(x_j^2)}&space;\cdot&space;L_{\lceil&space;\text{log}_2n&space;\rceil&space;-1}^{(x_j^2)}&space;\bigg(&space;\&space;.&space;\&space;.&space;\&space;.&space;\&space;.&space;\&space;.&space;\&space;.&space;\&space;\bigg)&space;R_{\lceil&space;\text{log}_2n&space;\rceil&space;-1}^{(x_j^{-2})}&space;\cdot&space;R_{\lceil&space;\text{log}_2n&space;\rceil}^{(x_j^{-2})}&space;\\&space;&=&space;\left(&space;\prod_{j=1}^{\lceil&space;\text{log}_2n&space;\rceil}&space;L_{j}^{(x_j^2)}&space;\right)&space;\cdot&space;P&space;\cdot&space;\left(&space;\prod_{j=1}^{\lceil&space;\text{log}_2n&space;\rceil}&space;R_{j}^{(x_j^{-2})}&space;\right)&space;\end{align}" title="\begin{align*} P_{\lceil \text{log}_2n \rceil + 1} &= L_{\lceil \text{log}_2n \rceil}^{(x_j^2)} \cdot P_{\lceil \text{log}_2n \rceil} \cdot R_{\lceil \text{log}_2n \rceil}^{(x_j^{-2})} \\ &= L_{\lceil \text{log}_2n \rceil}^{(x_j^2)} \cdot \left( L_{\lceil \text{log}_2n \rceil -1}^{(x_j^2)} \cdot P_{\lceil \text{log}_2n \rceil -1} \cdot R_{\lceil \text{log}_2n \rceil -1}^{(x_j^{-2})} \right) \cdot R_{\lceil \text{log}_2n \rceil}^{(x_j^{-2})} \\ &= L_{\lceil \text{log}_2n \rceil}^{(x_j^2)} \cdot L_{\lceil \text{log}_2n \rceil -1}^{(x_j^2)} \bigg( \ . \ . \ . \ . \ . \ . \ \bigg) R_{\lceil \text{log}_2n \rceil -1}^{(x_j^{-2})} \cdot R_{\lceil \text{log}_2n \rceil}^{(x_j^{-2})} \\ &= \left( \prod_{j=1}^{\lceil \text{log}_2n \rceil} L_{j}^{(x_j^2)} \right) \cdot P \cdot \left( \prod_{j=1}^{\lceil \text{log}_2n \rceil} R_{j}^{(x_j^{-2})} \right) \end{align}" /></a>

We can compute <a href="https://www.codecogs.com/eqnedit.php?latex=g,h" target="_blank"><img src="https://latex.codecogs.com/gif.latex?g,h" title="g,h" /></a> by unrolling the recursion and write them in the form:

<a href="https://www.codecogs.com/eqnedit.php?latex=\begin{align*}&space;g&space;&=&space;\prod_{i&space;=&space;1}^{n}&space;g_i^{s_i}&space;\in&space;\mathbb{G}&space;\\&space;h&space;&=&space;\prod_{i&space;=&space;1}^{n}&space;h_i^{s_i^{-1}}&space;\in&space;\mathbb{G}&space;\end{align*}" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\begin{align*}&space;g&space;&=&space;\prod_{i&space;=&space;1}^{n}&space;g_i^{s_i}&space;\in&space;\mathbb{G}&space;\\&space;h&space;&=&space;\prod_{i&space;=&space;1}^{n}&space;h_i^{s_i^{-1}}&space;\in&space;\mathbb{G}&space;\end{align*}" title="\begin{align*} g &= \prod_{i = 1}^{n} g_i^{s_i} \in \mathbb{G} \\ h &= \prod_{i = 1}^{n} h_i^{s_i^{-1}} \in \mathbb{G} \end{align*}" /></a>

Now to compute `s` vector <a href="https://www.codecogs.com/eqnedit.php?latex=(s_1,&space;s_2,&space;\dots,&space;s_n)" target="_blank"><img src="https://latex.codecogs.com/gif.latex?(s_1,&space;s_2,&space;\dots,&space;s_n)" title="(s_1, s_2, \dots, s_n)" /></a>, let's see it through an example for _n=8_.

![CodeCogsEqn](https://user-images.githubusercontent.com/19621621/83976087-68731000-a915-11ea-9818-cce27d18de0f.png)

![CodeCogsEqn (3)](https://user-images.githubusercontent.com/19621621/83978937-7a11e300-a928-11ea-83ef-5204b392ee2b.png)

NOTE that I intended to write <a href="https://www.codecogs.com/eqnedit.php?latex=\inline&space;x_j^{-1}" target="_blank"><img src="https://latex.codecogs.com/png.latex?\inline&space;x_j^{-1}" title="x_j^{-1}" /></a> but instead mistakenly wrote <a href="https://www.codecogs.com/eqnedit.php?latex=\inline&space;-x_j" target="_blank"><img src="https://latex.codecogs.com/png.latex?\inline&space;-x_j" title="-x_j" /></a>. Apologies for the same.

It's pretty clear why the exponents of <a href="https://www.codecogs.com/eqnedit.php?latex=\inline&space;g_i,&space;h_i" target="_blank"><img src="https://latex.codecogs.com/png.latex?\inline&space;g_i,&space;h_i" title="g_i, h_i" /></a> are the inverses of each other for all <a href="https://www.codecogs.com/eqnedit.php?latex=\inline&space;i=1,2,\dots,&space;n." target="_blank"><img src="https://latex.codecogs.com/png.latex?\inline&space;i=1,2,\dots,&space;n." title="i=1,2,\dots, n." /></a>. There is a pattern in the powers of triplet <a href="https://www.codecogs.com/eqnedit.php?latex=(x_1,x_2,x_3)" target="_blank"><img src="https://latex.codecogs.com/gif.latex?(x_1,x_2,x_3)" title="(x_1,x_2,x_3)" /></a> such that

<a href="https://www.codecogs.com/eqnedit.php?latex=s_i&space;=&space;\prod_{j=1}^{\lceil&space;\text{log}_2n&space;\rceil}&space;x_j^{b_{i,j}}&space;\text{&space;where&space;}&space;b_{i,j}&space;=&space;\begin{cases}&space;1&space;&&space;\text{if&space;}&space;j\text{-th&space;bit&space;of&space;}&space;i-1&space;\text{&space;is&space;}&space;1&space;\\&space;-1&space;&&space;\text{otherwise}&space;\end{cases}" target="_blank"><img src="https://latex.codecogs.com/gif.latex?s_i&space;=&space;\prod_{j=1}^{\lceil&space;\text{log}_2n&space;\rceil}&space;x_j^{b_{i,j}}&space;\text{&space;where&space;}&space;b_{i,j}&space;=&space;\begin{cases}&space;1&space;&&space;\text{if&space;}&space;j\text{-th&space;bit&space;of&space;}&space;i-1&space;\text{&space;is&space;}&space;1&space;\\&space;-1&space;&&space;\text{otherwise}&space;\end{cases}" title="s_i = \prod_{j=1}^{\lceil \text{log}_2n \rceil} x_j^{b_{i,j}} \text{ where } b_{i,j} = \begin{cases} 1 & \text{if } j\text{-th bit of } i-1 \text{ is } 1 \\ -1 & \text{otherwise} \end{cases}" /></a>

This part is a bit confusing but when we see our example again, it is much easy to understand.

<a href="https://www.codecogs.com/eqnedit.php?latex=\begin{align*}&space;i&space;=&space;1&space;&&space;\implies&space;\&space;(b_{1,1},&space;b_{1,2},&space;b_{1,3})&space;=&space;(-1,&space;-1,&space;-1)&space;&&space;\equiv&space;\&space;000&space;\\&space;i&space;=&space;2&space;&&space;\implies&space;\&space;(b_{2,1},&space;b_{2,2},&space;b_{2,3})&space;=&space;(-1,&space;-1,&space;1)&space;&&space;\equiv&space;\&space;001&space;\\&space;i&space;=&space;3&space;&&space;\implies&space;\&space;(b_{3,1},&space;b_{3,2},&space;b_{3,3})&space;=&space;(-1,&space;1,&space;-1)&space;&&space;\equiv&space;\&space;010&space;\\&space;\vdots&space;&&space;&&space;\\&space;i&space;=&space;8&space;&&space;\implies&space;\&space;(b_{8,1},&space;b_{8,2},&space;b_{8,3})&space;=&space;(1,&space;1,&space;1)&space;&&space;\equiv&space;\&space;111&space;\\&space;\end{align}" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\begin{align*}&space;i&space;=&space;1&space;&&space;\implies&space;\&space;(b_{1,1},&space;b_{1,2},&space;b_{1,3})&space;=&space;(-1,&space;-1,&space;-1)&space;&&space;\equiv&space;\&space;000&space;\\&space;i&space;=&space;2&space;&&space;\implies&space;\&space;(b_{2,1},&space;b_{2,2},&space;b_{2,3})&space;=&space;(-1,&space;-1,&space;1)&space;&&space;\equiv&space;\&space;001&space;\\&space;i&space;=&space;3&space;&&space;\implies&space;\&space;(b_{3,1},&space;b_{3,2},&space;b_{3,3})&space;=&space;(-1,&space;1,&space;-1)&space;&&space;\equiv&space;\&space;010&space;\\&space;\vdots&space;&&space;&&space;\\&space;i&space;=&space;8&space;&&space;\implies&space;\&space;(b_{8,1},&space;b_{8,2},&space;b_{8,3})&space;=&space;(1,&space;1,&space;1)&space;&&space;\equiv&space;\&space;111&space;\\&space;\end{align}" title="\begin{align*} i = 1 & \implies \ (b_{1,1}, b_{1,2}, b_{1,3}) = (-1, -1, -1) & \equiv \ 000 \\ i = 2 & \implies \ (b_{2,1}, b_{2,2}, b_{2,3}) = (-1, -1, 1) & \equiv \ 001 \\ i = 3 & \implies \ (b_{3,1}, b_{3,2}, b_{3,3}) = (-1, 1, -1) & \equiv \ 010 \\ \vdots & & \\ i = 8 & \implies \ (b_{8,1}, b_{8,2}, b_{8,3}) = (1, 1, 1) & \equiv \ 111 \\ \end{align}" /></a>

Therefore, for given <a href="https://www.codecogs.com/eqnedit.php?latex=(i,j)" target="_blank"><img src="https://latex.codecogs.com/gif.latex?(i,j)" title="(i,j)" /></a>, we can easily compute vector component `s_i`.

## Implementation Idea

First we compute <a href="https://www.codecogs.com/eqnedit.php?latex=s_1&space;=&space;x_k^{-1}&space;\cdot&space;x_{k-1}^{-1}&space;\dots&space;x_2^{-1}&space;\cdot&space;x_1^{-1}&space;\text{&space;where&space;}&space;k=\lceil&space;\text{log}_2n&space;\rceil" target="_blank"><img src="https://latex.codecogs.com/gif.latex?s_1&space;=&space;x_k^{-1}&space;\cdot&space;x_{k-1}^{-1}&space;\dots&space;x_2^{-1}&space;\cdot&space;x_1^{-1}&space;\text{&space;where&space;}&space;k=\lceil&space;\text{log}_2n&space;\rceil" title="s_1 = x_k^{-1} \cdot x_{k-1}^{-1} \dots x_2^{-1} \cdot x_1^{-1} \text{ where } k=\lceil \text{log}_2n \rceil" /></a>. Note the order of the challenges is reverse since challenges are stored in order of creation as <a href="https://www.codecogs.com/eqnedit.php?latex=[x_k,&space;x_{k-1},&space;\dots,&space;x_1]" target="_blank"><img src="https://latex.codecogs.com/gif.latex?[x_k,&space;x_{k-1},&space;\dots,&space;x_1]" title="[x_k, x_{k-1}, \dots, x_1]" /></a>.

Now, we for next values in `s` vector:

<a href="https://www.codecogs.com/eqnedit.php?latex=\begin{align*}&space;s_2&space;&=&space;s_1&space;\cdot&space;x_3^{2}&space;\\&space;s_3&space;&=&space;s_1&space;\cdot&space;x_{2}^{2}&space;\\&space;s_4&space;&=&space;s_2&space;\cdot&space;x_{2}^2&space;\\&space;\vdots&space;\end{align}" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\begin{align*}&space;s_2&space;&=&space;s_1&space;\cdot&space;x_3^{2}&space;\\&space;s_3&space;&=&space;s_1&space;\cdot&space;x_{2}^{2}&space;\\&space;s_4&space;&=&space;s_2&space;\cdot&space;x_{2}^2&space;\\&space;\vdots&space;\end{align}" title="\begin{align*} s_2 &= s_1 \cdot x_3^{2} \\ s_3 &= s_1 \cdot x_{2}^{2} \\ s_4 &= s_2 \cdot x_{2}^2 \\ \vdots \end{align}" /></a>.

The logic is that for computing <a href="https://www.codecogs.com/eqnedit.php?latex=s_i" target="_blank"><img src="https://latex.codecogs.com/gif.latex?s_i" title="s_i" /></a> we need some <a href="https://www.codecogs.com/eqnedit.php?latex=s_{i-k}" target="_blank"><img src="https://latex.codecogs.com/gif.latex?s_{i-k}" title="s_{i-k}" /></a> in which we just need to flip _one_ bit to get the desired configuration of <a href="https://www.codecogs.com/eqnedit.php?latex=s_i" target="_blank"><img src="https://latex.codecogs.com/gif.latex?s_i" title="s_i" /></a>. Let me know if there are any issues in my explanation.

Reply:
Please note there could also be an alternate way of computing <a href="https://www.codecogs.com/eqnedit.php?latex=s_i" target="_blank"><img src="https://latex.codecogs.com/gif.latex?s_i" title="s_i" /></a>. Let us discuss if we can do it an easier way.

Reply:
Got it! thank you for the thorough explanation! neat idea !